mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp

```diff
@@ -0,0 +1,789 @@
+//===- AffineLoopInterchange.cpp - Pass to perform loop interchange-----===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements a loop interchange pass that optimizes for locality
+// (spatial and temporal - both self and group) and parallelism for multicores,
+// to minimize the frequency of synchronization. The pass works for both
+// perfectly nested and imperfectly nested loops (any level of nesting). However
+// in the presence of affine.if statements and/or non-rectangular iteration
+// space, the pass simply bails out - leaving the original loop nest unchanged.
+// The pass is triggered by the command line flag -affine-loop-interchange.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetail.h"
+#include "mlir/Analysis/AffineAnalysis.h"
+#include "mlir/Analysis/Utils.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Affine/Passes.h"
+#include "mlir/Transforms/LoopUtils.h"
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/ADT/SmallVector.h"
+#include <algorithm>
+#include <cmath>
+#include <numeric>
+using namespace mlir;
+namespace {
+struct LoopInterchange : public AffineLoopInterchangeBase<LoopInterchange> {
+  void runOnFunction() override;
+  void handleImperfectlyNestedAffineLoops(Operation &funcOp);
+  void runOnAffineLoopNest();
+
+private:
+  /// Default cache line size(in bytes). Useful for getting a measure of the
+  /// locality of each loop in a given loop nest.
+  constexpr static unsigned kCacheLineSize = 64;
+
+  /// Default element size to be used if a memref does not have a static shape.
+  constexpr static unsigned kDefaultEltSize = 8;
```

Comment:
Dynamic shape doesn't mean the element size is unknown. The elt size is always known as long as it's an int/float type. So this comment looks incorrect.